src/metabase/driver/mongo/query_processor.clj

```diff
@@ -211,16 +234,45 @@
 
 ;;; ### filter
 
+;; omf: 'max' date in a date range is next day at 00:00:00.000
+(defn- resolve-between [basetype minvalue maxvalue]
+  (cond (= basetype :type/DateTime)
+    {$gte minvalue
+     $lt (t/plus maxvalue (t/days 1))}
+  :else
+    {$gte minvalue
+     $lte maxvalue})
+)
+
+;; omf: a single day ends before next day at 00:00:00.000
+(defn- resolve-equal [basetype value unit]
+  (cond (= basetype :type/DateTime)
+    {$gte value
+     $lt (case unit
+           :year (t/plus value (t/years 1))
+           :month (t/plus value (t/months 1))
+           :week (t/plus value (t/weeks 1))
+           (t/plus value (t/days 1)))}
+  :else
+    {"$eq" value})
```

Comment:
Try to take advantage of the fact that conditional logic can happen anywhere in Lisp languages, e.g. you could have done:

```clojure
(if (isa? basetype :type/DateTime)
  {$gte value
   $lt  (t/plus value (case unit
                        :year  (t/years 1)
                        :month (t/months 1)
                        :week  (t/weeks 1)
                        (t/days 1)))}
  {$eq value})
```